### PR TITLE
Add new API methods

### PR DESCRIPTION
**Pull request by 333444555**
Has been tested, and seems to be working, test further for confirmation.
Adds HTTP API methods with documentation:
new_samples -> POST /samples
get_samples -> GET /samples
new_app -> POST /app
get_apps -> GET /apps
get_app -> GET /apps/.app-id
update_app -> PUT /app/.app-id
delete_app -> DELETE /app/.app-id

### CODE REVIEW

**Comment by patapizza** [comment]
Thanks for the pull request! We'll review it shortly.

File: wit/wit.py [hunk 1]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
             params['context'] = json.dumps(context)
         resp = req(self.logger, self.access_token, 'GET', '/message', params)
         return resp
+    def new_samples(self, text,entities,*args):
```
Comment (patapizza) [inline wit/wit.py]: * add empty line before
* rename to `post_samples`
* space after each comma

File: wit/wit.py [hunk 2]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        """  Validate samples (sentence + entities annotations) to train your app programmatically.
```
Comment (patapizza) [inline wit/wit.py]: new line after """
Just link to the http doc as a single source of truth: https://wit.ai/docs/http (remove all docstring until the list of params.

File: wit/wit.py [hunk 3]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+          }]
+          
+          
+        :param text: The text (sentence) you want your app to understand.
+        :param entities: The list of entities appearing in this sentence, that you want your app to extract once it is trained. (use the format above)
+        
+        :return: Dictionary that contains if the samples were sent ['sent'], and number of samples sent ['n']"""
+        params = {'v': WIT_API_VERSION}
+        data = []
+        if text and entities:
```
Comment (patapizza) [inline wit/wit.py]: if not, throw

File: wit/wit.py [hunk 4]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+            data.append({'text': text,'entities': entities})
```
Comment (patapizza) [inline wit/wit.py]: space after ,

File: wit/wit.py [hunk 5]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        if len(args)>1:
```
Comment (patapizza) [inline wit/wit.py]: stay simple: remove this block and get rid of parameter `args`

File: wit/wit.py [hunk 6]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+            x=0
+            while x<len(args):
+                data.append({'text': args[x],'entities': args[x+1]})
+                x+=2
+        resp = req(self.logger, self.access_token, 'POST', '/samples', params,data=json.dumps(data))
```
Comment (patapizza) [inline wit/wit.py]: inline, return directly no need for `resp`

File: wit/wit.py [hunk 7]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
             params['context'] = json.dumps(context)
         resp = req(self.logger, self.access_token, 'GET', '/message', params)
```
Comment (patapizza) [inline wit/wit.py]: we can return directly here too

File: wit/wit.py [hunk 8]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+            data.append({'text': text,'entities': entities})
+        if len(args)>1:
+            x=0
+            while x<len(args):
+                data.append({'text': args[x],'entities': args[x+1]})
+                x+=2
+        resp = req(self.logger, self.access_token, 'POST', '/samples', params,data=json.dumps(data))
+        return resp
+    def get_samples(self, limit, offset=None, entity_ids=None, entity_values=None, negative=None):
+        """Return a list of Samples.
```
Comment (patapizza) [inline wit/wit.py]: lint to http docs

File: wit/wit.py [hunk 9]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        if text and entities:
+            data.append({'text': text,'entities': entities})
+        if len(args)>1:
+            x=0
+            while x<len(args):
+                data.append({'text': args[x],'entities': args[x+1]})
+                x+=2
+        resp = req(self.logger, self.access_token, 'POST', '/samples', params,data=json.dumps(data))
+        return resp
+    def get_samples(self, limit, offset=None, entity_ids=None, entity_values=None, negative=None):
```
Comment (patapizza) [inline wit/wit.py]: add empty line between functions

File: wit/wit.py [hunk 10]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        
+        :param limit: Max number of samples to return. Must be between 1 and 10000 inclusive.
+        :param offset: Number of samples to skip. Must be >= 0. Default is 0.
+        :param entity_ids: List of names of the entities that all returned samples must include. 
+        :param entity_values: List of values of the specified entities that all returned samples must include. Note entity_values can only be specified if entity_ids is specified.
+        :param negative: A flag to show the negative samples of the specified entity_id. 
+        
+        :return: a list of all samples."""
+        params = {}
+        if limit:
```
Comment (patapizza) [inline wit/wit.py]: if not, throw, required

File: wit/wit.py [hunk 11]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+            params['limit'] = limit
+        if offset:
+            params['offset'] = offset
+        if entity_ids:
+            params['entity_ids'] = entity_ids
+        if entity_values:
+            params['entity_values'] = negative
```
Comment (patapizza) [inline wit/wit.py]: `params['entity_values'] = ','.join(entity_values)`

File: wit/wit.py [hunk 12]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        :param negative: A flag to show the negative samples of the specified entity_id. 
+        
+        :return: a list of all samples."""
+        params = {}
+        if limit:
+            params['limit'] = limit
+        if offset:
+            params['offset'] = offset
+        if entity_ids:
+            params['entity_ids'] = entity_ids
```
Comment (patapizza) [inline wit/wit.py]: `params['entity_ids'] = ','.join(entity_ids)`

File: wit/wit.py [hunk 13]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        if entity_values:
+            params['entity_values'] = negative
+        if negative:
+            params['negative'] = negative
+        resp = req(self.logger, self.access_token, 'GET', '/samples', params)
```
Comment (patapizza) [inline wit/wit.py]: return directly

File: wit/wit.py [hunk 14]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        return resp
+    def new_app(self, name, lang, private, desc=None):
```
Comment (patapizza) [inline wit/wit.py]: newline
rename to `post_apps`

File: wit/wit.py [hunk 15]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        """Creates a new app for an existing user.
+        :param name: Name of the new app.
+        :param lang: Language code (ISO alpha-2 code) e.g : "en". 
+        :param private: Private if “true”
+        
+        :return: A dictionary with the new app's access token ['access_token'] and the new app's id ['app_id']"""
+        params = {'v': WIT_API_VERSION}
+        data = None
+        if name and lang and private:
```
Comment (patapizza) [inline wit/wit.py]: if not throw

File: wit/wit.py [hunk 16]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        if entity_ids:
+            params['entity_ids'] = entity_ids
+        if entity_values:
+            params['entity_values'] = negative
+        if negative:
+            params['negative'] = negative
+        resp = req(self.logger, self.access_token, 'GET', '/samples', params)
+        return resp
+    def new_app(self, name, lang, private, desc=None):
+        """Creates a new app for an existing user.
```
Comment (patapizza) [inline wit/wit.py]: link to doc

File: wit/wit.py [hunk 17]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        :param private: Private if “true”
+        
+        :return: A dictionary with the new app's access token ['access_token'] and the new app's id ['app_id']"""
+        params = {'v': WIT_API_VERSION}
+        data = None
+        if name and lang and private:
+            data = {'name': name,'lang': lang,'private': str(private).lower()}
+        if desc:
+            data['desc'] = desc
+        resp = req(self.logger, self.access_token, 'POST', '/apps', params,data=json.dumps(data))
```
Comment (patapizza) [inline wit/wit.py]: return directly

File: wit/wit.py [hunk 18]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        return resp
+    def get_apps(self, limit, offset=None):
```
Comment (patapizza) [inline wit/wit.py]: newline

File: wit/wit.py [hunk 19]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        """Returns an array of all apps that you own.
```
Comment (patapizza) [inline wit/wit.py]: link to doc

File: wit/wit.py [hunk 20]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        NOTE: Does not return shared apps.
+        
+        :param limit: Max number of apps to return. Must be between 1 and 10000 inclusive (recommended max size is 500).
+        :param offset: Number of apps to skip. Must be >= 0. Default is 0.
+        
+        :return: List of all apps with their information."""
+        params = {'v': WIT_API_VERSION}
+        if limit:
```
Comment (patapizza) [inline wit/wit.py]: if not throw

File: wit/wit.py [hunk 21]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+            params['limit'] = limit
+        if offset:
+            params['offset'] = offset
+        resp = req(self.logger, self.access_token, 'GET', '/apps', params)
```
Comment (patapizza) [inline wit/wit.py]: ditto

File: wit/wit.py [hunk 22]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        return resp
+    def update_app(self, app_id, name=None, lang=None, private=None, timezone=None, desc=None):
```
Comment (patapizza) [inline wit/wit.py]: newline, rename to `put_apps`

File: wit/wit.py [hunk 23]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        """Updates an app with the given attributes.
```
Comment (patapizza) [inline wit/wit.py]: link to doc

File: wit/wit.py [hunk 24]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+            data['name'] = name
+        if lang:
+            data['lang'] = lang
+        if private:
+            data['private'] = private
+        if timezone:
+            data['timezone'] = timezone
+        if desc:
+            data['desc'] = desc
+        resp = req(self.logger, self.access_token, 'PUT', '/apps/'+app_id, params,data=json.dumps(data))
```
Comment (patapizza) [inline wit/wit.py]: space around `+`
if data is empty, throw, no-op
return directly

File: wit/wit.py [hunk 25]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        return resp
+    def get_app(self, app_id):
+        """Returns a map representation of the specified app.
+        
+        :param app_id: The ID of the app you want to information about.
+        
+        :return: Information about the requested app."""
+        resp = req(self.logger, self.access_token, 'GET', '/apps/'+app_id, {})
```
Comment (patapizza) [inline wit/wit.py]: return directly
space around `+`

File: wit/wit.py [hunk 26]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        if private:
+            data['private'] = private
+        if timezone:
+            data['timezone'] = timezone
+        if desc:
+            data['desc'] = desc
+        resp = req(self.logger, self.access_token, 'PUT', '/apps/'+app_id, params,data=json.dumps(data))
+        return resp
+    def get_app(self, app_id):
+        """Returns a map representation of the specified app.
```
Comment (patapizza) [inline wit/wit.py]: link to doc

File: wit/wit.py [hunk 27]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        
+        :param app_id: The ID of the app you want to information about.
+        
+        :return: Information about the requested app."""
+        resp = req(self.logger, self.access_token, 'GET', '/apps/'+app_id, {})
+        return resp
+    def delete_app(self, app_id):
```
Comment (patapizza) [inline wit/wit.py]: newline

File: wit/wit.py [hunk 28]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        """Permanently delete the app.
```
Comment (patapizza) [inline wit/wit.py]: link to doc

File: wit/wit.py [hunk 29]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        NOTE: You must be the creator of the app to be able to delete it.
+        
+        :param app_id: ID of the want-to-be deleted app.
+        
+        :return: Dictionary with a boolean that represents the success of the operation ['success']."""
+        resp = req(self.logger, self.access_token, 'DELETE', '/apps/'+app_id, {})
```
Comment (patapizza) [inline wit/wit.py]: return directly, space around `+`

File: wit/wit.py [hunk 30]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        return resp
+    def speech(self, audio_file, verbose=None, headers=None):
+        """ Sends an audio file to the /speech API.
+        Uses the streaming feature of requests (see `req`), so opening the file
+        in binary mode is strongly reccomended (see
+        http://docs.python-requests.org/en/master/user/advanced/#streaming-uploads).
+        Add Content-Type header as specified here: https://wit.ai/docs/http/20160526#post--speech-link
```
Comment (patapizza) [inline wit/wit.py]: link to latest doc (https://wit.ai/docs/http)

File: wit/wit.py [hunk 31]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+
+        :param audio_file: an open handler to an audio file
+        :param verbose:
+        :param headers: an optional dictionary with request headers
+        :return:
+        """
+        params = {}
+        headers = headers or {}
```
Comment (patapizza) [inline wit/wit.py]: move `{}` to param definition

File: wit/wit.py [hunk 32]
```diff
@@ -69,6 +69,151 @@ def message(self, msg, context=None, n=None, verbose=None):
...
+        if verbose:
+            params['verbose'] = True
+        resp = req(self.logger, self.access_token, 'POST', '/speech', params,
```
Comment (patapizza) [inline wit/wit.py]: return directly

**Comment by cpngwn** [comment]
are you able submit edits?

**Pull request closed without merging**